tools/nvqpp-v1/nvqpp_driver.cpp

```diff
@@ -0,0 +1,685 @@
+ * All rights reserved.                                                        *
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ ******************************************************************************/
+#include "nvqpp_driver.h"
+#include "cudaq/Frontend/nvqpp/ASTBridge.h"
+#include "cudaq/Optimizer/Dialect/CC/CCDialect.h"
+#include "cudaq/Optimizer/Dialect/Quake/QuakeDialect.h"
+#include "cudaq/Optimizer/Support/Verifier.h"
+#include "frontend_actions.h"
+#include "nvqpp_flag_configs.h"
+#include "nvqpp_options.h"
+#include "clang/CodeGen/BackendUtil.h"
+#include "clang/CodeGen/CodeGenAction.h"
+#include "clang/Driver/Compilation.h"
+#include "clang/Driver/DriverDiagnostic.h"
+#include "clang/Driver/Job.h"
+#include "clang/Frontend/CompilerInstance.h"
+#include "clang/Frontend/FrontendAction.h"
+#include "clang/Frontend/FrontendActions.h"
+#include "clang/Tooling/Tooling.h"
+#include "llvm/Support/BuryPointer.h"
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/ErrorOr.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/InitLLVM.h"
+#include "llvm/Support/MemoryBuffer.h"
+#include "llvm/Support/Process.h"
+#include "llvm/Support/Program.h"
+#include "llvm/Support/Signals.h"
+#include "llvm/Support/TargetSelect.h"
+#include "llvm/Support/TimeProfiler.h"
+#include "llvm/Support/Timer.h"
+#include "llvm/Support/ToolOutputFile.h"
+#include "llvm/Support/VirtualFileSystem.h"
+#include "llvm/TargetParser/Host.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/InitAllPasses.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassManager.h"
+
+extern "C" {
+void getThisExecutablePath() { return; }
+}
+
+namespace cudaq {
+Driver::Driver(ArgvStorageBase &cmdArgs)
+    : driverPath(llvm::sys::fs::getMainExecutable(
+          cmdArgs[0], (void *)(intptr_t)getThisExecutablePath)),
+      diag(cmdArgs, driverPath),
+      clangDriver(driverPath, llvm::sys::getDefaultTargetTriple(), diag.engine,
+                  "nvq++ compiler") {
+  clangDriver.ResourceDir = std::string(CLANG_RESOURCE_DIR);
+  setInstallDir(cmdArgs);
+  std::tie(clOptions, hostCompilerArgs) = preProcessCudaQArguments(cmdArgs);
+  if (clOptions.hasArg(cudaq::nvqpp::options::OPT_target)) {
+    const std::string clTargetName =
+        clOptions.getLastArgValue(cudaq::nvqpp::options::OPT_target, "").str();
+    auto target = TargetRegistry::lookupTarget(clTargetName);
+    if (target.has_value()) {
+      targetName = clTargetName;
+      targetPlatformExtraArgs =
+          target->generateConfig(clOptions, cudaqTargetsPath);
+
+      // Insert backend config to the list of files to compile
+      const std::string backendConfigCppFile =
+          cudaqTargetsPath + "/backendConfig.cpp";
```

Comment:
I do recall that Clang / LLVM had a FileSystem abstraction, and in the past I've used one of its specializations to write a "file" to the current program memory space, without ever needing to actually output a file to disk. I wonder if we could use that here and avoid having the `backendConfig.cpp` file, and just use a string template defined in this file.